src/derive.rs

```diff
@@ -282,16 +282,16 @@ pub trait Subcommand: FromArgMatches + Sized {
 /// }
 /// ```
 pub trait ArgEnum: Sized {
-    /// All possible argument choices, in display order.
-    const VARIANTS: &'static [&'static str];
+    /// All possible argument values, in display order.
+    fn arg_values() -> Vec<ArgValue<'static>>;
```

Comment:
Sorry, I missed this `Vec` on my last pass.   This is requiring an allocation for possible values.  It is only once and not much data is being copied, so its not the worst, but it'd still be nice to avoid.

Can't use `[]` because a size is needed.

I know `&[]` is annoying but it seems like it'd give best results and the annoying part is limited to one piece of code we generate.

Reply:
Thanks for the reminder about alias.

The only workaround I can come up with is if we instead provide the value variants (function or constant).  The user still pays for the cost for aliases, if they use them, but no other allocations would be made.   This also opens us up to providing an inherent method for `from_str`.

What are your thoughts?

Reply:
I'm not understanding what your proposing.
could you write down the signatures you are proposing?

Reply:
```rust
pub trait ArgEnum: Sized {
    fn value_variants() -> &[Self];

    fn from_str(input: &str, case_insensitive: bool) -> Result<Self, String> {
        Self::value_variants().into_iter().find(|v| v.arg_value().unwrap().matches(input, case_insensitive)).ok_or_else(...)
    }
    fn arg_value(&self) -> Option<ArgValue<'static>>;
}
```

We can then have
```rust
fn gen_arg_enum_possible_values(ty: &Type) -> TokenStream {
    quote_spanned! { ty.span()=>
        .possible_values(<#ty as clap::ArgEnum>::value_variants().into_iter().map(clap::ArgEnum::arg_value))
    }
}
```
(of course, slight tweaks might be needed to handle references vs owned types)

Reply:
This will have problems with & vs no &:
```
error[E0308]: mismatched types
   --> src/derive.rs:290:9
    |
284 |  / pub trait ArgEnum: Sized + 'static {
285 |  |     /// All possible argument values, in display order.
286 |  |     fn value_variants() -> &'static [Self];
287 |  |
288 |  |     /// Parse an argument into `Self`.
289 |  |     fn from_str(input: &str, case_insensitive: bool) -> Result<Self, String> {
    |  |                                                         -------------------- expected `std::result::Result<Self, String>` because of return type
290 | /|         Self::value_variants()
291 | ||             .into_iter()
292 | ||             .find(|v| v.arg_value().unwrap().matches(input, case_insensitive))
293 | ||             .ok_or_else(|| format!("Invalid variant: {}", input))
    | ||_________________________________________________________________^ expected type parameter `Self`, found `&Self`
...    |
299 |  |     fn arg_value(&self) -> Option<ArgValue<'static>>;
300 |  | }
    |  |_- this type parameter
    |
    = note: expected enum `std::result::Result<Self, _>`
               found enum `std::result::Result<&Self, _>`

For more information about this error, try `rustc --explain E0308`.
```

We either need to call cloned/copied or return Result<&Self>

Reply:
**If** we do this, I lean towards `Result<&Self>`.  Its ok for the proc macro to assume copy/clone and it is **probably** fine for the `trait` to make `Copy` or `Clone` a super trait, but I'm concerned with what people might put in skipped variants.  We'd want to at most do `Clone` and not `Copy` so people can have something like a `String` in it.  Returning `&Self` though gives us full flexibility at the cost of ergonomics.

So the question is ergonomics (`Result<Self>`) is worth the cost of flexibility (requiring `Clone`).

Reply:
I feel like I'm running against a brick wall with this trait definition and lifetimes etc. maybe I'll come back  on your ofter to speak on Discord

Reply:
I'd recommend getting this into a working state, even if not ideal, and we merge it as-is.  We can then play with it and meet up on platform-of-choice and discuss trade offs.

Reply:
Alright I have pushed a solution using a slice but requiring Clone.